Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -446,4 +451,716 @@ alias nonsingular_of_affine_of_Z_ne_zero := nonsingular_of_Z_ne_zero
+    + 4 * W'.a₂ * W'.a₃ * W'.a₄ * P x * P z ^ 3 - 2 * W'.a₃ ^ 3 * P x * P z ^ 3
+    + W'.a₃ * W'.a₄ ^ 2 * P z ^ 4 - W'.a₁ ^ 4 * P x ^ 2 * P y * P z
+    + W'.a₁ ^ 3 * W'.a₂ * P x ^ 3 * P z - 2 * W'.a₁ ^ 3 * W'.a₃ * P x * P y * P z ^ 2
+    + W'.a₁ ^ 3 * W'.a₄ * P x ^ 2 * P z ^ 2 + W'.a₁ ^ 2 * W'.a₂ * W'.a₃ * P x ^ 2 * P z ^ 2
+    - W'.a₁ ^ 2 * W'.a₃ ^ 2 * P y * P z ^ 3 + 2 * W'.a₁ ^ 2 * W'.a₃ * W'.a₄ * P x * P z ^ 3
+    - W'.a₁ * W'.a₂ * W'.a₃ ^ 2 * P x * P z ^ 3 - W'.a₂ * W'.a₃ ^ 3 * P z ^ 4
+    + W'.a₁ * W'.a₃ ^ 2 * W'.a₄ * P z ^ 4
+
+lemma dblX_eq' {P : Fin 3 → R} (hP : W'.Equation P) : W'.dblX P * P z =
+    (eval P W'.polynomialX ^ 2 - W'.a₁ * eval P W'.polynomialX * P z * (P y - W'.negY P)
+      - W'.a₂ * P z ^ 2 * (P y - W'.negY P) ^ 2 - 2 * P x * P z * (P y - W'.negY P) ^ 2)
+      * (P y - W'.negY P) := by
+  linear_combination (norm := (rw [dblX, eval_polynomialX, negY]; ring1))
+    9 * (W'.a₁ * P x ^ 2 + 2 * P x * P y) * (equation_iff _).mp hP
+
+lemma dblX_eq {P : Fin 3 → F} (hP : W.Equation P) (hPz : P z ≠ 0) : W.dblX P =
+    ((eval P W.polynomialX ^ 2 - W.a₁ * eval P W.polynomialX * P z * (P y - W.negY P)
+      - W.a₂ * P z ^ 2 * (P y - W.negY P) ^ 2 - 2 * P x * P z * (P y - W.negY P) ^ 2)
+      * (P y - W.negY P)) / P z := by
+  rw [← dblX_eq' hP, mul_div_cancel_right₀ _ hPz]
+
+lemma dblX_smul (P : Fin 3 → R) (u : R) : W'.dblX (u • P) = u ^ 4 * W'.dblX P := by
+  simp only [dblX, smul_fin3_ext]
+  ring1
+
+lemma dblX_of_Z_eq_zero [NoZeroDivisors R] {P : Fin 3 → R} (hP : W'.Equation P) (hPz : P z = 0) :
+    W'.dblX P = 0 := by
+  rw [dblX, hPz, X_eq_zero_of_Z_eq_zero hP hPz]
+  ring1
+
+lemma dblX_of_Y_eq [NoZeroDivisors R] {P Q : Fin 3 → R} (hP : W'.Equation P) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z) (hy : P y * Q z = Q y * P z)
+    (hy' : P y * Q z = W'.negY Q * P z) : W'.dblX P = 0 := by
+  apply eq_zero_of_ne_zero_of_mul_right_eq_zero hPz
+  rw [dblX_eq' hP, Y_eq_negY_of_Y_eq hQz hx hy hy']
+  ring1
+
+private lemma toAffine_addX_of_eq {P : Fin 3 → F} (hPz : P z ≠ 0) {n d : F} (hd : d ≠ 0) :
+    W.toAffine.addX (P x / P z) (P x / P z) (-n / P z / d) =
+      (n ^ 2 - W.a₁ * n * P z * d - W.a₂ * P z ^ 2 * d ^ 2 - 2 * P x * P z * d ^ 2) * d / P z
+        / (P z * d ^ 3) := by
+  field_simp [mul_ne_zero hPz hd]
+  ring1
+
+lemma dblX_of_Z_ne_zero {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z = Q x * P z) (hy : P y * Q z ≠ W.negY Q * P z) :
+    W.dblX P / W.dblZ P = W.toAffine.addX (P x / P z) (Q x / Q z)
+      (W.toAffine.slope (P x / P z) (Q x / Q z) (P y / P z) (Q y / Q z)) := by
+  rw [dblX_eq hP hPz, dblZ, toAffine_slope_of_eq hP hQ hPz hQz hx hy, ← (X_eq_iff hPz hQz).mp hx,
+    toAffine_addX_of_eq hPz <| sub_ne_zero.mpr <| Y_ne_negY_of_Y_ne' hP hQ hPz hQz hx hy]
+
+variable (W') in
+/-- The $Y$-coordinate of a representative of `-(2 • P)` for a point `P`. -/
+noncomputable def negDblY (P : Fin 3 → R) : R :=
```

Comment:
Note to myself: 
`z^2  (negDblY - y polynomialY^3 - c1 polynomialY) + polynomialX^3 = c2 polynomial`,
where
```
c1 = -3  a1^2  x^2  y - 12 a1 x y^2 - 18 y^3 - a1 a2^2 x^2 z + 
 3 a1 a4 x^2 z - 4 a2^2 x y z - 6 a1 a3 x y z + 12 a4 x y z + 
 a1^3 y^2 z + 5 a1 a2 y^2 z - 27 a3 y^2 z - 2 a2^2 a3 x z^2 + 
 6 a3 a4 x z^2 + 4 a1 a2 a3 y z^2 - 9 a3^2 y z^2 - 2 a1^2 a4 y z^2 - 
 2 a2 a4 y z^2 + 18 a6 y z^2 - a2 a3 a4 z^3 + a1 a4^2 z^3 - 
 3 a1 a2 a6 z^3 + 9 a3  a6  z^3
c2 = 3 (9 x^3 + 9 a2 x^2 z + 9 y^2 z - a1^2 a2 x z^2 + 3 a1 a3 x z^2 - 
   2 a1 a2 y z^2 + 12 a3 y z^2 - a1 a2 a3 z^3 + 3 a3^2 z^3)
```

Reply:
Better: the cubic factor in `dblX` can be replaced by 
```
-a2 x^3 + a1 x^2 y + x y^2 - a1 a3 x^2 z - 3 a4 x^2 z + a3 x y z - 
 2 a3^2 x z^2 - 9 a6 x z^2 - a2 a3^2 z^3 + a1 a3 a4 z^3 + a4^2 z^3 - 
 a1^2 a6 z^3 - 4 a2 a6 z^3
```
which is
`x polynomialY^2 - ψh₃ - 3 (x/z) polynomial`.
Now trying to do the same for `negDblY` ...

Reply:
Originally I worked out this formula in order to show when you apply `negDblY` to a 2-torsion nonsingular point (`polynomialY = 0` but `polynomialX ≠ 0`) then `negDblY ≠ 0`, so our doubling formula still holds in the P+P=O case. Now, with my accidental discovery that `negDblY` can be taken to be `Y polynomialY^3 - ψh₃ polynomialX` we can give an alternative proof: at a 2-torsion nonsingular point P ≠ O we have `ψh₃ ≠ 0` because otherwise P is also 3-torsion and therefore P = O; since polynomialY=0 and polynomialX≠0 we conclude negDblY≠0. This is a bit circular though, as it uses a high-powered result #13782, even though the PR uses only Jacobian API which is independent of the projective case we consider here.